### ISSUE TITLE
Feature: Support private Maven repositories

### ISSUE THREAD

**Issue by davidgoate** [dependabot/feedback#229]
Sometimes we use internal private dependencies that are not published to Maven Central. For example, we use a hosted jFrog Artifactory.

Our `pom.xml` would have 

```xml
<repositories>
        <repository>
            <id>our-org-name</id>
            <name>some-name</name>
            <url>https://our-org.jfrog.io/our-org/our-repo</url>
        </repository>
    </repositories>
```

and we would reference some dependencies that are resolved to that repository in the `pom.xml` `<dependencies/>`

Is there "some way" (as if by magic) that dependabot can work with these? I presume ultimately besides the possibly complex question of authentication, these are really just spec compliant maven repos so in theory there is a chance that whatever process dependabot uses on Maven central to find new releases could also work on private repos? Perhaps an org wide dashboard setting can be given for the the encrypted maven credential (much the same as we would give our CI or our developers to put in their local `~/.m2/settings.xml` file, e.g.:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<settings xsi:schemaLocation="http://maven.apache.org/SETTINGS/1.1.0 http://maven.apache.org/xsd/settings-1.1.0.xsd" xmlns="http://maven.apache.org/SETTINGS/1.1.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <servers>
    <server>
      <username>some-user</username>
      <password>abc-123</password>
      <id>our-org-name</id>
    </server>
  </servers>
</settings>
```

P.S. we do follow semver for publishing artifacts here.

**Comment by RohanNagar** [comment]
Private repos are supported if you go to the Dependabot dashboard. In the top right, click on your username, then click on "Config Variables". From there you can add a Maven repository.

**Comment by davidgoate** [comment]
@RohanNagar thanks for letting me know. Is there a way I can debug this?

What I've just done:

1. Create a new user in artifactory.
1. Add my repo in the config variables
1. Manually triggered a bump of pom.xml in a couple of repos that depend on packages which have newer semver versions published in my private repo
1. Waited until bumping was done
1. Observed no new PR's

My first thought are, is there an authentication or permissions error followed by, how does dependabot make the link between my `pom.xml` 

```xml
<repositories>
        <repository>
            <id>our-org-name</id>
            <name>some-name</name>
            <url>https://our-org.jfrog.io/our-org/our-repo</url>
        </repository>
    </repositories>
``` and the config variable? Is it base on the URL - unlike the `~/.m2/settings.xml` which seems to use the `id`

**Comment by RohanNagar** [comment]
Not sure about the internals here - someone on the Dependabot team will have to answer 😄

**Comment by davidgoate** [comment]
@RohanNagar I have found an access log in artifactory which shows access denied errors for the user.  The strange thing is I can login via the UI ok with that password. I've tried both the UI password and encrypted password token that the jFrog UI generates for maven auth. I'll keep digging. thanks for your help

**Comment by davidgoate** [comment]
In case anypne else finds into this, my issue was that it seems that artifactory doesn't allow BASIC auth using the login password or encrypted password.

I solved this by finding [this](https://www.jfrog.com/confluence/display/RTF/Access+Tokens#AccessTokens-BasicAuthentication) which suggested I [create a token ](https://www.jfrog.com/confluence/display/RTF/Artifactory+REST+API#ArtifactoryRESTAPI-CreateToken).

After running a command like this as admin:

```bash
curl -u admin:my-pass -XPOST "https://my-org.jfrog.io/my-org/api/security/token" -d "username=dependabot" -d "scope=mmber-of-groups:backend-developers" -d "expires_in=0"
{
  "scope" : "member-of-groups:backend-developers api:*",
  "access_token" : "the-token",
  "token_type" : "Bearer"
}
```

Then I was able to use the `access_token` in the config variables and finally the process worked. I'm still not really clear why I **can** create an access token using HTTP BASIC AUTH with my encrypted password but **I can't** provide dependabot with the encrypted password to have it do the bumping.

To be clear, when I talk about encrypted password and password, I am referring to these things in jFrog on the profile page of the user:
  
![screenshot 2018-10-24 at 22 12 14](https://user-images.githubusercontent.com/8789912/47461937-ebdffc00-d7d9-11e8-90d8-e21bc034b614.png)

**Comment by greysteil** [comment]
Sorry for the delay picking this up - just getting to it now.

Dependabot does indeed scan your `pom.xml` to figure out all of the registries it might want to use. It then authenticates using basic auth in requests like [this one](https://github.com/dependabot/dependabot-core/blob/master/lib/dependabot/update_checkers/java/maven/version_finder.rb#L111-L117).

It looks like you've got things working using a token, but I'd like to get to the bottom of why any other setup wasn't working, if you think it should have been.
- The code Dependabot uses to associate auth details and credentials is [here](https://github.com/dependabot/dependabot-core/blob/master/lib/dependabot/update_checkers/java/maven/version_finder.rb#L128-L131). As you can see, it basically constructs a separate array of registries to try from your config variables (rather than trying to combine things together).
- That means that if you enter the `username` and `password` for a Dependabot config variable exactly as you would in `~/.m2/settings.xml` (for the `username` and `password` attributes) everything should work

If the above *isn't* working then I definitely want to dig into it!

**Comment by davidgoate** [comment]
@greysteil everything's been working for the last couple of days. What I noticed before was that when I added the password as the "encrypted password" as jfrog calls it (the same one I would add to my `~/.m2/settings.xml`) nothing would happen in dependabot when i triggered a bumb and in the artifactory logs I would see:

> 2018-10-24 20:35:48,874 [DENIED LOGIN]  for dependabot/54.91.51.58. 

in the `access.log`


What I ended up finding after googling for a while was:

>An access token can be used instead of a password for basic authentication. This may be useful when you need a client (such as certain dependency managers) that only supports basic authentication to access Artifactory. In this case, it is important to access Artifactory using the same user name provided when creating the token

On [this page](https://www.jfrog.com/confluence/display/RTF/Access+Tokens#AccessTokens-BasicAuthentication). Which made me go down the route of creating an access token. After this, it all started working.  I'm not really clear why the encrypted password is ok for an `mvn clean install` but not for authentication using these API's - but it does all seem to work ok for me now.
